File: backend/app/api/v1/schemas/categories.py
from __future__ import annotations

from typing import List
from pydantic import BaseModel, ConfigDict, Field


# =========================================================
# Category Output
# =========================================================

class CategoryOut(BaseModel):
    """
    Public category representation.

    Used by:
        • /categories
        • UI filters
        • onboarding flows
    """

    model_config = ConfigDict(
        from_attributes=True,
        frozen=True,
    )

    id: str
    name: str = Field(..., min_length=1)

    icon: str | None = None
    color: str | None = None


# =========================================================
# Categories Response
# =========================================================

class CategoriesResponse(BaseModel):
    """
    Category collection response.

    Deterministic ordering handled upstream.
    """

    model_config = ConfigDict(
        frozen=True,
    )

    total: int = Field(..., ge=0)

    items: List[CategoryOut] = Field(default_factory=list)

File: frontend/src/api/categories.ts
import { client } from './client';

export interface CategoryOut {
  id: string;
  name: string;
  icon: string | null;
  color: string | null;
}

export interface CategoriesResponse {
  total: number;
  items: CategoryOut[];
}

export async function fetchCategories(): Promise<CategoryOut[]> {
  const { data } = await client.get<CategoriesResponse>('/api/v1/categories');
  if (!Array.isArray(data?.items)) return [];
  return data.items.filter((c) => c.id && c.name);
}
